File: app/src/main/java/com/prey/actions/triggers/TriggerReceiver.java
/*******************************************************************************
 * Created by Orlando Aliaga
 * Copyright 2019 Prey Inc. All rights reserved.
 * License: GPLv3
 * Full license at "/LICENSE"
 ******************************************************************************/
package com.prey.actions.triggers;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;

import com.prey.PreyConfig;
import com.prey.PreyLogger;
import com.prey.actions.observer.ActionResult;
import com.prey.util.ClassUtil;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

public abstract class TriggerReceiver extends BroadcastReceiver {

    public abstract void onReceive(Context context, Intent intent);

    public void execute(Context context, String name) {
        TriggerDataSource dataSource = new TriggerDataSource(context);
        List<TriggerDto> listTrigger = dataSource.getAllTriggers();
        PreyLogger.d("Trigger TriggerReceiver onReceive name:" + name);
        PreyLogger.d("Trigger TriggerReceiver onReceive listTrigger.size():" + (listTrigger == null ? -1 : listTrigger.size()));
        for (int i = 0; listTrigger != null && i < listTrigger.size(); i++) {
            TriggerDto trigger = listTrigger.get(i);
            List<TriggerEventDto> listEvents = TriggerParse.TriggerEvents(trigger.getEvents());
            String triggerName = trigger.getName();
            PreyLogger.d("Trigger TriggerReceiver triggerName:" + triggerName);
            for (int j = 0; listEvents != null && j < listEvents.size(); j++) {
                TriggerEventDto event = listEvents.get(j);
                if (name.equals(event.type)) {
                    boolean process = true;
                    boolean haveRange = TriggerUtil.haveRange(listEvents);
                    if (haveRange) {
                        PreyLogger.d("Trigger TriggerReceiver  haveRange:" + haveRange);
                        boolean validRange = TriggerUtil.validRange(listEvents);
                        PreyLogger.d("Trigger TriggerReceiver  validRange:" + validRange);
                        process = validRange;
                    }
                    try {
                        if (process) {
                            PreyLogger.d("Trigger TriggerReceiver triggerName trigger.getActions():" + trigger.getActions());
                            executeActions(context, trigger.getActions());
                        }
                    } catch (Exception e) {
                        PreyLogger.e("e:" + e.getMessage(), e);
                    }
                }
            }
        }
    }

    public void executeActions(final Context context, String actions) throws Exception {
        List<TriggerActionDto> listActions = TriggerParse.TriggerActions(actions);
        for (int z = 0; listActions != null && z < listActions.size(); z++) {
            final TriggerActionDto actionDto = listActions.get(z);
            int delay = actionDto.delay;
            PreyLogger.d("triggerName TriggerReceiver delay:" + delay);
            if (delay > 0) {
                Thread.sleep(delay * 1000);
            }
            new Thread() {
                public void run() {
                    try {
                        PreyLogger.d("Trigger triggerName actionDto.action:" + actionDto.action);
                        JSONObject jsonObject = new JSONObject(actionDto.action);
                        PreyLogger.d("Trigger triggerName action:" + jsonObject);
                        PreyLogger.d("Trigger triggerName jsonObject:" + jsonObject);
                        String nameAction = jsonObject.getString("target");
                        PreyLogger.d("Trigger triggerName nameAction:" + nameAction);
                        String methodAction = jsonObject.getString("command");
                        PreyLogger.d("Trigger triggerName methodAction:" + methodAction);
                        JSONObject parametersAction = null;
                        try {
                            parametersAction = jsonObject.getJSONObject("options");
                            PreyLogger.d("Trigger triggerName parametersAction:" + parametersAction);
                        } catch (JSONException e) {
                        }
                        if (parametersAction == null) {
                            parametersAction = new JSONObject();
                        }
                        try {
                            String messageId = jsonObject.getString(PreyConfig.MESSAGE_ID);
                            parametersAction.put(PreyConfig.MESSAGE_ID, messageId);
                        } catch (Exception e) {
                        }
                        PreyLogger.d("Trigger nameAction:" + nameAction + " methodAction:" + methodAction + " parametersAction:" + parametersAction);
                        List<ActionResult> lista = new ArrayList<ActionResult>();
                        ClassUtil.execute(context, lista, nameAction, methodAction, parametersAction, null);
                    } catch (Exception e) {
                        PreyLogger.e("Trigger error:" + e.getMessage(), e);
                    }
                }
            }.start();
        }
    }
}

File: app/src/main/java/com/prey/actions/triggers/TriggerReceiver.kt
/*******************************************************************************
 * Created by Orlando Aliaga
 * Copyright 2025 Prey Inc. All rights reserved.
 * License: GPLv3
 * Full license at "/LICENSE"
 ******************************************************************************/
package com.prey.actions.triggers

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent

import com.prey.actions.observer.ActionResult
import com.prey.PreyLogger
import com.prey.util.ClassUtil
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

import org.json.JSONException
import org.json.JSONObject

/**
 * Abstract base class for trigger receivers.
 * Provides a basic implementation for executing triggers and actions.
 */
abstract class TriggerReceiver : BroadcastReceiver() {
    abstract override fun onReceive(context: Context, intent: Intent)

    /**
     * Executes a trigger with the given name.
     * Retrieves all triggers from the data source, checks if the trigger has a matching event,
     * and executes the corresponding actions.
     *
     * @param context The Context in which the receiver is running.
     * @param name The name of the trigger to execute.
     */
    fun execute(context: Context, name: String) {
        val dataSource = TriggerDataSource(context)
        val listTrigger = dataSource.allTriggers
        PreyLogger.d("Trigger TriggerReceiver onReceive name:$name")
        PreyLogger.d(
            "Trigger TriggerReceiver onReceive listTrigger.size():${
                (listTrigger.size
                    ?: -1)
            }"
        )
        var i = 0
        while (listTrigger != null && i < listTrigger.size) {
            val trigger = listTrigger[i]
            val listEvents = TriggerParse.TriggerEvents(trigger.getEvents())
            var j = 0
            while (listEvents != null && j < listEvents.size) {
                val event = listEvents[j]
                PreyLogger.d("Trigger TriggerReceiver onReceive name:${name} event.type:${event.getType()}")
                if (name == event.getType()) {
                    var process = true
                    val haveRange = TriggerUtil.haveRange(listEvents)
                    PreyLogger.d("Trigger TriggerReceiver  haveRange:$haveRange")
                    if (haveRange) {
                        val validRange = TriggerUtil.validRange(listEvents)
                        PreyLogger.d("Trigger TriggerReceiver  validRange:$validRange")
                        process = validRange
                    }
                    try {
                        if (process) {
                            PreyLogger.d("Trigger TriggerReceiver triggerName trigger.getActions():${trigger.getActions()}")
                            executeActions(context, trigger.getActions())
                        }
                    } catch (e: Exception) {
                        PreyLogger.e("Error:${e.message}", e)
                    }
                }
                j++
            }
            i++
        }
    }

    /**
     * Executes a list of actions.
     * Parses the actions from the given string, executes each action with the given delay,
     * and logs any errors that occur.
     *
     * @param context The Context in which the receiver is running.
     * @param actions The string containing the actions to execute.
     * @throws Exception If an error occurs while executing the actions.
     */
    @Throws(Exception::class)
    fun executeActions(context: Context, actions: String) {
        val listActions = TriggerParse.TriggerActions(actions)
        var z = 0
        while (listActions != null && z < listActions.size) {
            val actionDto = listActions[z]
            val delay = actionDto.getDelay()
            PreyLogger.d("triggerName TriggerReceiver delay:$delay")
            if (delay > 0) {
                Thread.sleep((delay * 1000).toLong())
            }
            CoroutineScope(Dispatchers.IO).launch {
                try {
                    PreyLogger.d("Trigger triggerName actionDto.action:${actionDto.getAction()}")
                    val jsonObject = JSONObject(actionDto.getAction())
                    PreyLogger.d("Trigger triggerName action:$jsonObject")
                    PreyLogger.d("Trigger triggerName jsonObject:$jsonObject")
                    val nameAction = jsonObject.getString("target")
                    PreyLogger.d("Trigger triggerName nameAction:$nameAction")
                    val methodAction = jsonObject.getString("command")
                    PreyLogger.d("Trigger triggerName methodAction:$methodAction")
                    var parametersAction: JSONObject? = null
                    try {
                        parametersAction = jsonObject.getJSONObject("options")
                        PreyLogger.d("Trigger triggerName parametersAction:$parametersAction")
                    } catch (e: JSONException) {
                        PreyLogger.e("Error:${e.message}", e)
                    }
                    if (parametersAction == null) {
                        parametersAction = JSONObject()
                    }
                    PreyLogger.d("Trigger nameAction:$nameAction methodAction:$methodAction parametersAction:$parametersAction")
                    val listActions: MutableList<ActionResult> = ArrayList()
                    ClassUtil.getInstance().execute(
                        context,
                        listActions,
                        nameAction,
                        methodAction,
                        parametersAction,
                        null
                    )
                } catch (e: Exception) {
                    PreyLogger.e("Trigger error:${e.message}", e)
                }
            }
            z++
        }
    }

}